sketch/Explorer_ESP_WiFi-localization_v1.2.ino

```diff
@@ -0,0 +1,162 @@
+
+// TTN constructor
+TheThingsNetwork ttn(loraSerial, debugSerial, freqPlan);
+
+void setup()
+{
+  // Wait a maximum of 10s for Serial Monitor
+  while (!debugSerial && millis() < 10000)
+    ;
+
+  loraSerial.begin(57600);
+  debugSerial.begin(ESP_SERIAL_SPEED);
+  espSerial.begin(ESP_SERIAL_SPEED);
+
+  debugSerial.println("-- STATUS");
+  ttn.showStatus();
+
+  debugSerial.println("-- JOIN");
+  ttn.join(appEui, appKey);
+}
+
+void loop()
+{
+  uint8_t bssidNumber = 0;
+  uint32_t timeout;
+  // command to esp, set station mode
+  espSerial.println(F("AT+CWMODE=1"));
+  bool x = waitForOK();
+
+  if (!x)
+  {
+    debugSerial.println(F("Set mode failed"));
+    goto exit;
+  }
+
+  // Command to esp to get accesspoint information
+  espSerial.println(F("AT+CWLAP"));
+
+  char *line;
+  timeout = millis() + 3000;
+
+  while (millis() < timeout && bssidNumber < ACCESS_POINTS)
+  {
+    line = readLine();
+    debugSerial.println(line);
+    if (strncmp("OK", line, 2) == 0)
+    {
+      debugSerial.println(F("end of access points"));
+      break;
+    }
+    if (strncmp("+CWLAP", line, 2) != 0)
+    {
+      delay(1);
+      debugSerial.println(F("No Bssid"));
+      continue;
+    }
+
+    // Break the accesspoint in tokens, to get the bssid out.
+    strtok(line, "\"");
+    strtok(NULL, "\"");
+    strtok(NULL, "\"");
+    char *bssid = strtok(NULL, "\"");
+
+    // Check if a valid bssid string has been found
+    if (bssid && strlen(bssid) == SF_WIFI_BSSID_SIZE * 3 - 1)
+    {
+      debugSerial.print(F("Bssid: "));
+      debugSerial.println(bssid);
+
+      // putting the bssid into an array for easy access
+      char *b = strtok(bssid, ":");
+      for (bssidByte = 0; bssidByte < SF_WIFI_BSSID_SIZE && b; bssidByte++, b = strtok(NULL, ":"))
+      {
+        aps[bssidNumber][bssidByte] = HEX_PAIR_TO_BYTE(b[0], b[1]);
+      }
+      bssidNumber++;
+    }
+  }
+
+  sendBssid();
+exit:
+  debugSerial.println("Restart the loop");
+}
+
+bool waitForOK(uint32_t waitTime)
+{
+  uint32_t timeout = millis() + waitTime;
+  while (millis() < timeout)
+  {
+    uint8_t l = espSerial.readBytesUntil('\n', buffer, SF_ESP8266_BUFFER_SIZE);
+    if (l > 0 && strncmp("OK", buffer, 2) == 0)
+    {
+      return true;
+    }
+    delay(1);
+  }
+  return false;
+}
+
+char *readLine(uint32_t waitTime)
+{
+  uint32_t timeout = millis() + waitTime;
+  while (millis() < timeout)
+  {
+    uint8_t l = espSerial.readBytesUntil('\n', buffer, SF_ESP8266_BUFFER_SIZE);
+    if (l > 0)
+    {
+      buffer[l - 1] = '\0';
+      return buffer;
+    }
+    delay(1);
+  }
+  return NULL;
+}
+
+void sendBssid()
+{
+  byte payload[payloadLength];
+  uint8_t payloadByte = 0;
+  for (bssidNumber = 0; bssidNumber < ACCESS_POINTS; bssidNumber++)
+  {
+    for (bssidByte = 0; bssidByte < SF_WIFI_BSSID_SIZE; bssidByte++, payloadByte++)
+    {
+      payload[payloadByte] = aps[bssidNumber][bssidByte];
+    }
+  }
+  ttn.sendBytes(payload, sizeof(payload));
+  delay(300000);
```

Comment:
Do not put the delay here, but in the loop. Be clear about timing in the loop, not in some utility function

Reply:
Also make this a define or const